File: trunk/ase/awk/StdAwk.cpp
/*
 * $Id: StdAwk.cpp,v 1.29 2007/11/08 15:08:06 bacon Exp $
 *
 * {License}
 */

#include <ase/awk/StdAwk.hpp>
#include <ase/cmn/str.h>
#include <ase/utl/stdio.h>
#include <ase/utl/ctype.h>

#include <stdlib.h>
#include <math.h>
#include <time.h>

#ifdef _WIN32
	#include <tchar.h>
#else
	#include <wchar.h>
#endif

/////////////////////////////////
ASE_BEGIN_NAMESPACE(ASE)
/////////////////////////////////


StdAwk::StdAwk ()
{
	seed = (unsigned int)::time(NULL);
	::srand (seed);
}

#define ADD_FUNC(name,min,max,impl) \
	do { \
		if (addFunction (name, min, max, \
			(FunctionHandler)impl) == -1)  \
		{ \
			Awk::close (); \
			return -1; \
		} \
	} while (0)

int StdAwk::open ()
{
	int n = Awk::open ();
	if (n == -1) return n;

	ADD_FUNC (ASE_T("sin"),        1, 1, &StdAwk::sin);
	ADD_FUNC (ASE_T("cos"),        1, 1, &StdAwk::cos);
	ADD_FUNC (ASE_T("tan"),        1, 1, &StdAwk::tan);
	ADD_FUNC (ASE_T("atan"),       1, 1, &StdAwk::atan);
	ADD_FUNC (ASE_T("atan2"),      2, 2, &StdAwk::atan2);
	ADD_FUNC (ASE_T("log"),        1, 1, &StdAwk::log);
	ADD_FUNC (ASE_T("exp"),        1, 1, &StdAwk::exp);
	ADD_FUNC (ASE_T("sqrt"),       1, 1, &StdAwk::sqrt);
	ADD_FUNC (ASE_T("int"),        1, 1, &StdAwk::fnint);
	ADD_FUNC (ASE_T("rand"),       0, 0, &StdAwk::rand);
	ADD_FUNC (ASE_T("srand"),      0, 1, &StdAwk::srand);
	ADD_FUNC (ASE_T("systime"),    0, 0, &StdAwk::systime);
	ADD_FUNC (ASE_T("strftime"),   0, 2, &StdAwk::strftime);
	ADD_FUNC (ASE_T("strfgmtime"), 0, 2, &StdAwk::strfgmtime);
	ADD_FUNC (ASE_T("system"),     1, 1, &StdAwk::system);

	return 0;
}

int StdAwk::sin (Run& run, Return& ret, const Argument* args, size_t nargs, 
	const char_t* name, size_t len)
{
	return ret.set ((real_t)::sin(args[0].toReal()));
}

int StdAwk::cos (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	return ret.set ((real_t)::cos(args[0].toReal()));
}

int StdAwk::tan (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	return ret.set ((real_t)::tan(args[0].toReal()));
}

int StdAwk::atan (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	return ret.set ((real_t)::atan(args[0].toReal()));
}

int StdAwk::atan2 (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	return ret.set ((real_t)::atan2(args[0].toReal(), args[1].toReal()));
}

int StdAwk::log (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	return ret.set ((real_t)::log(args[0].toReal()));
}

int StdAwk::exp (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	return ret.set ((real_t)::exp(args[0].toReal()));
}

int StdAwk::sqrt (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	return ret.set ((real_t)::sqrt(args[0].toReal()));
}

int StdAwk::fnint (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	return ret.set (args[0].toInt());
}

int StdAwk::rand (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	return ret.set ((long_t)::rand());
}

int StdAwk::srand (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	unsigned int prevSeed = seed;

	seed = (nargs == 0)? 
		(unsigned int)::time(NULL):
		(unsigned int)args[0].toInt();
	::srand (seed);
	return ret.set ((long_t)prevSeed);
}

#if defined(_WIN32) && defined(_MSC_VER) && (_MSC_VER>=1400)
	#define time_t __time64_t
	#define time _time64
	#define localtime _localtime64
	#define gmtime _gmtime64
#endif

int StdAwk::systime (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	return ret.set ((long_t)::time(NULL));
}

int StdAwk::strftime (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	const char_t* fmt;
	size_t fln;
       
	fmt = (nargs < 1)? ASE_T("%c"): args[0].toStr(&fln);
	time_t t = (nargs < 2)? ::time(NULL): (time_t)args[1].toInt();

	char_t buf[128]; 
	struct tm* tm;
#ifdef _WIN32
	tm = ::localtime (&t);
#else
	struct tm tmb;
	tm = ::localtime_r (&t, &tmb);
#endif

#ifdef ASE_CHAR_IS_MCHAR
	size_t l = ::strftime (buf, ASE_COUNTOF(buf), fmt, tm);
#else
	size_t l = ::wcsftime (buf, ASE_COUNTOF(buf), fmt, tm);
#endif

	return ret.set (buf, l);	
}

int StdAwk::strfgmtime (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	const char_t* fmt;
	size_t fln;
       
	fmt = (nargs < 1)? ASE_T("%c"): args[0].toStr(&fln);
	time_t t = (nargs < 2)? ::time(NULL): (time_t)args[1].toInt();

	char_t buf[128]; 
	struct tm* tm;
#ifdef _WIN32
	tm = ::gmtime (&t);
#else
	struct tm tmb;
	tm = ::gmtime_r (&t, &tmb);
#endif

#ifdef ASE_CHAR_IS_MCHAR
	size_t l = ::strftime (buf, ASE_COUNTOF(buf), fmt, tm);
#else
	size_t l = ::wcsftime (buf, ASE_COUNTOF(buf), fmt, tm);
#endif

	return ret.set (buf, l);	
}

int StdAwk::system (Run& run, Return& ret, const Argument* args, size_t nargs,
	const char_t* name, size_t len)
{
	size_t l;
	const char_t* ptr = args[0].toStr(&l);

#ifdef _WIN32
	return ret.set ((long_t)::_tsystem(ptr));
#elif defined(ASE_CHAR_IS_MCHAR)
	return ret.set ((long_t)::system(ptr));
#else
	char* mbs = (char*)ase_awk_malloc (awk, l*5+1);
	if (mbs == ASE_NULL) return -1;

	::size_t mbl = ::wcstombs (mbs, ptr, l*5);
	if (mbl == (::size_t)-1) 
	{
		ase_awk_free (awk, mbs);
		return -1;
	}
	mbs[mbl] = '\0';
	int n = ret.set ((long_t)::system(mbs));

	ase_awk_free (awk, mbs);
	return n;
#endif
}

int StdAwk::openPipe (Pipe& io) 
{ 
	Awk::Pipe::Mode mode = io.getMode();
	FILE* fp = NULL;

	switch (mode)
	{
		case Awk::Pipe::READ:
			fp = ase_popen (io.getName(), ASE_T("r"));
			break;
		case Awk::Pipe::WRITE:
			fp = ase_popen (io.getName(), ASE_T("w"));
			break;
	}

	if (fp == NULL) return -1;

	io.setHandle (fp);
	return 1;
}

int StdAwk::closePipe (Pipe& io) 
{
	fclose ((FILE*)io.getHandle());
	return 0; 
}

StdAwk::ssize_t StdAwk::readPipe (Pipe& io, char_t* buf, size_t len) 
{ 
	FILE* fp = (FILE*)io.getHandle();
	ssize_t n = 0;

	while (n < (ssize_t)len)
	{
		ase_cint_t c = ase_fgetc (fp);
		if (c == ASE_CHAR_EOF) break;

		buf[n++] = c;
		if (c == ASE_T('\n')) break;
	}

	return n;
}

StdAwk::ssize_t StdAwk::writePipe (Pipe& io, char_t* buf, size_t len) 
{ 
	FILE* fp = (FILE*)io.getHandle();
	size_t left = len;

	while (left > 0)
	{
		if (*buf == ASE_T('\0')) 
		{
		#if defined(ASE_CHAR_IS_WCHAR) && defined(__linux)
			if (fputc ('\0', fp) == EOF)
		#else
			if (ase_fputc (*buf, fp) == ASE_CHAR_EOF) 
		#endif
			{
				return -1;
			}
			left -= 1; buf += 1;
		}
		else
		{
		#if defined(ASE_CHAR_IS_WCHAR) && defined(__linux)
		// fwprintf seems to return an error with the file
		// pointer opened by popen, as of this writing. 
		// anyway, hopefully the following replacement 
		// will work all the way.
			int chunk = (left > ASE_TYPE_MAX(int))? ASE_TYPE_MAX(int): (int)left;	
			int n = fprintf (fp, "%.*ls", chunk, buf);
			if (n >= 0)
			{
				size_t x;
				for (x = 0; x < chunk; x++)
				{
					if (buf[x] == ASE_T('\0')) break;
				}
				n = x;
			}
		#else
			int chunk = (left > ASE_TYPE_MAX(int))? ASE_TYPE_MAX(int): (int)left;
			int n = ase_fprintf (fp, ASE_T("%.*s"), chunk, buf);
		#endif

			if (n < 0 || n > chunk) return -1;
			left -= n; buf += n;
		}
	}

	return len;
}

int StdAwk::flushPipe (Pipe& io) 
{ 
	return ::fflush ((FILE*)io.getHandle()); 
}

// file io handlers 
int StdAwk::openFile (File& io) 
{ 
	Awk::File::Mode mode = io.getMode();
	FILE* fp = NULL;

	switch (mode)
	{
		case Awk::File::READ:
			fp = ase_fopen (io.getName(), ASE_T("r"));
			break;
		case Awk::File::WRITE:
			fp = ase_fopen (io.getName(), ASE_T("w"));
			break;
		case Awk::File::APPEND:
			fp = ase_fopen (io.getName(), ASE_T("a"));
			break;
	}

	if (fp == NULL) return -1;

	io.setHandle (fp);
	return 1;
}

int StdAwk::closeFile (File& io) 
{ 
	fclose ((FILE*)io.getHandle());
	return 0; 
}

StdAwk::ssize_t StdAwk::readFile (File& io, char_t* buf, size_t len) 
{
	FILE* fp = (FILE*)io.getHandle();
	ssize_t n = 0;

	while (n < (ssize_t)len)
	{
		ase_cint_t c = ase_fgetc (fp);
		if (c == ASE_CHAR_EOF) break;

		buf[n++] = c;
		if (c == ASE_T('\n')) break;
	}

	return n;
}

StdAwk::ssize_t StdAwk::writeFile (File& io, char_t* buf, size_t len)
{
	FILE* fp = (FILE*)io.getHandle();
	size_t left = len;

	while (left > 0)
	{
		if (*buf == ASE_T('\0')) 
		{
			if (ase_fputc (*buf, fp) == ASE_CHAR_EOF) return -1;
			left -= 1; buf += 1;
		}
		else
		{
			int chunk = (left > ASE_TYPE_MAX(int))? ASE_TYPE_MAX(int): (int)left;
			int n = ase_fprintf (fp, ASE_T("%.*s"), chunk, buf);
			if (n < 0 || n > chunk) return -1;
			left -= n; buf += n;
		}
	}

	return len;
}

int StdAwk::flushFile (File& io) 
{ 
	return ::fflush ((FILE*)io.getHandle()); 
}

// memory allocation primitives
void* StdAwk::allocMem (size_t n) 
{ 
	return ::malloc (n); 
}

void* StdAwk::reallocMem (void* ptr, size_t n) 
{ 
	return ::realloc (ptr, n); 
}

void  StdAwk::freeMem (void* ptr) 
{ 
	::free (ptr); 
}

// character class primitives
StdAwk::bool_t StdAwk::isUpper (cint_t c)
{ 
	return ase_isupper (c); 
}

StdAwk::bool_t StdAwk::isLower (cint_t c) 
{
	return ase_islower (c); 
}

StdAwk::bool_t StdAwk::isAlpha (cint_t c) 
{ 
	return ase_isalpha (c); 
}

StdAwk::bool_t StdAwk::isDigit (cint_t c) 
{
	return ase_isdigit (c); 
}

StdAwk::bool_t StdAwk::isXdigit (cint_t c)
{
	return ase_isxdigit (c); 
}

StdAwk::bool_t StdAwk::isAlnum (cint_t c) 
{
	return ase_isalnum (c); 
}

StdAwk::bool_t StdAwk::isSpace (cint_t c) 
{
	return ase_isspace (c); 
}

StdAwk::bool_t StdAwk::isPrint (cint_t c) 
{
	return ase_isprint (c); 
}

StdAwk::bool_t StdAwk::isGraph (cint_t c) 
{ 
	return ase_isgraph (c); 
}

StdAwk::bool_t StdAwk::isCntrl (cint_t c) 
{ 
	return ase_iscntrl (c); 
}

StdAwk::bool_t StdAwk::isPunct (cint_t c) 
{ 
	return ase_ispunct (c); 
}

StdAwk::cint_t StdAwk::toUpper (cint_t c) 
{ 
	return ase_toupper (c); 
}

StdAwk::cint_t StdAwk::toLower (cint_t c) 
{ 
	return ase_tolower (c); 
}

// miscellaneous primitive
StdAwk::real_t StdAwk::pow (real_t x, real_t y) 
{ 
	return ::pow (x, y); 
}

int StdAwk::vsprintf (
	char_t* buf, size_t size, const char_t* fmt, va_list arg) 
{
	return ase_vsprintf (buf, size, fmt, arg);
}

void StdAwk::vdprintf (const char_t* fmt, va_list arg) 
{
	ase_vfprintf (stderr, fmt, arg);
}

/////////////////////////////////
ASE_END_NAMESPACE(ASE)
/////////////////////////////////

File: trunk/ase/awk/StdAwk.java
/*
 * $Id: StdAwk.java,v 1.21 2007/11/10 15:30:07 bacon Exp $
 *
 * {License}
 */

package ase.awk;

import java.io.*;

/**
 * Extends the core interpreter engine to implement the language closer to 
 * the standard.
 */
public abstract class StdAwk extends Awk
{
	private long seed;
	private java.util.Random random;

	public StdAwk () throws Exception
	{
		super ();

		seed = System.currentTimeMillis();
		random = new java.util.Random (seed);

		addFunction ("sin", 1, 1); 
		addFunction ("cos", 1, 1); 
		addFunction ("tan", 1, 1); 
		addFunction ("atan", 1, 1); 
		addFunction ("atan2", 2, 2); 
		addFunction ("log", 1, 1); 
		addFunction ("exp", 1, 1); 
		addFunction ("sqrt", 1, 1); 
		addFunction ("int", 1, 1, "bfnint"); 

		addFunction ("srand", 0, 1); 
		addFunction ("rand", 0, 0); 

		addFunction ("systime", 0, 0);
		addFunction ("strftime", 0, 2);
		addFunction ("strfgmtime", 0, 2);

		addFunction ("system", 1, 1); 
	}

	/* == file interface == */
	protected int openFile (File file)
	{
		int mode = file.getMode();

		if (mode == File.MODE_READ)
		{
			FileInputStream fis;

			try { fis = new FileInputStream (file.getName()); }
			catch (IOException e) { return -1; }

			Reader rd = new BufferedReader (
				new InputStreamReader (fis)); 
			file.setHandle (rd);
			return 1;
		}
		else if (mode == File.MODE_WRITE)
		{
			FileOutputStream fos;

			try { fos = new FileOutputStream (file.getName()); }
			catch (IOException e) { return -1; }

			Writer wr = new BufferedWriter (
				new OutputStreamWriter (fos));
			file.setHandle (wr);
			return 1;
		}
		else if (mode == File.MODE_APPEND)
		{
			FileOutputStream fos;

			try { fos = new FileOutputStream (file.getName(), true); }
			catch (IOException e) { return -1; }

			Writer wr = new BufferedWriter (
				new OutputStreamWriter (fos));
			file.setHandle (wr);
			return 1;
		}

		return -1;
	}
	
	protected int closeFile (File file)
	{
		int mode = file.getMode();

		if (mode == File.MODE_READ)
		{
			Reader isr = (Reader)file.getHandle();
			try { isr.close (); }
			catch (IOException e) { return -1; }
			return 0;
		}
		else if (mode == File.MODE_WRITE)
		{
			Writer osw = (Writer)file.getHandle();
			try { osw.close (); }
			catch (IOException e) { return -1; }
			return 0;
		}
		else if (mode == File.MODE_APPEND)
		{
			Writer osw = (Writer)file.getHandle();
			try { osw.close (); }
			catch (IOException e) { return -1; }
			return 0;
		}
		
		return -1;
	}

	protected int readFile (File file, char[] buf, int len) 
	{
		int mode = file.getMode();

		if (mode == File.MODE_READ)
		{
			Reader rd = (Reader)file.getHandle();

			try 
			{
				len = rd.read (buf, 0, len);
				if (len == -1) return 0;
			}
			catch (IOException e) { return -1; }
			return len; 
		}

		return -1;
	}

	protected int writeFile (File file, char[] buf, int len) 
	{
		int mode = file.getMode();

		if (mode == File.MODE_WRITE ||
		    mode == File.MODE_APPEND)
		{
			Writer wr = (Writer)file.getHandle();
			try { wr.write (buf, 0, len); }
			catch (IOException e) { return -1; }
			return len;
		}

		return -1;
	}

	protected int flushFile (File file)
	{
		int mode = file.getMode ();

		if (mode == File.MODE_WRITE ||
		    mode == File.MODE_APPEND)
		{
			Writer wr = (Writer)file.getHandle ();
			try { wr.flush (); }
			catch (IOException e) { return -1; }
			return 0;
		}

		return -1;
	}

	private class RWE
	{
		public Writer wr;
		public Reader rd;
		public Reader er;

		public RWE (Writer wr, Reader rd, Reader er)
		{
			this.wr = wr;
			this.rd = rd;
			this.er = er;
		}
	};

	/* == pipe interface == */
	protected int openPipe (Pipe pipe)
	{
		int mode = pipe.getMode();

		if (mode == Pipe.MODE_READ)
		{
			Process proc;
		       
			try { proc = popen (pipe.getName()); }
			catch (IOException e) { return -1; }

			Reader rd = new BufferedReader (
				new InputStreamReader (proc.getInputStream())); 

			pipe.setHandle (rd);
			return 1;
		}
		else if (mode == Pipe.MODE_WRITE)
		{
			Process proc;

			try { proc = popen (pipe.getName()); }
			catch (IOException e) { return -1; }

			Writer wr = new BufferedWriter (
				new OutputStreamWriter (proc.getOutputStream()));
			Reader rd = new BufferedReader (
				new InputStreamReader (proc.getInputStream()));
			Reader er = new BufferedReader (
				new InputStreamReader (proc.getErrorStream()));

			pipe.setHandle (new RWE (wr, rd, er));
			return 1;
		}

		return -1;
	}
	
	protected int closePipe (Pipe pipe)
	{
		int mode = pipe.getMode();

		if (mode == Pipe.MODE_READ)
		{
			Reader rd = (Reader)pipe.getHandle();
			try { rd.close (); }
			catch (IOException e) { return -1; }
			return 0;
		}
		else if (mode == Pipe.MODE_WRITE)
		{
			//Writer wr = (Writer)pipe.getHandle();
			RWE rwe = (RWE)pipe.getHandle();

			try { rwe.wr.close (); }
			catch (IOException e) { return -1; }

			char[] buf = new char[256];

			try 
			{
				while (true)
				{
					int len = rwe.rd.read (buf, 0, buf.length);
					if (len == -1) break;
					System.out.print (new String (buf, 0, len));
				}

				System.out.flush ();
			}
			catch (IOException e) {}

			try
			{
				while (true)
				{
					int len = rwe.er.read (buf, 0, buf.length);
					if (len == -1) break;
					System.err.print (new String (buf, 0, len));
				}

				System.err.flush ();
			}
			catch (IOException e) {}

			try { rwe.rd.close (); } catch (IOException e) {}
			try { rwe.er.close (); } catch (IOException e) {}

			pipe.setHandle (null);
			return 0;
		}
		
		return -1;
	}

	protected int readPipe (Pipe pipe, char[] buf, int len) 
	{
		int mode = pipe.getMode();

		if (mode == Pipe.MODE_READ)
		{
			Reader rd = (Reader)pipe.getHandle();
			try 
			{
				len = rd.read (buf, 0, len);
				if (len == -1) len = 0;
			}
			catch (IOException e) { len = -1; }
			return len; 
		}

		return -1;
	}

	protected int writePipe (Pipe pipe, char[] buf, int len) 
	{
		int mode = pipe.getMode();

		if (mode == Pipe.MODE_WRITE)
		{
			//Writer wr = (Writer)pipe.getHandle ();
			RWE rw = (RWE)pipe.getHandle();
			try 
			{ 
				rw.wr.write (buf, 0, len); 
				rw.wr.flush ();
			}
			catch (IOException e) { return -1; }
			return len;
		}

		return -1;
	}

	protected int flushPipe (Pipe pipe)
	{
		int mode = pipe.getMode ();

		if (mode == Pipe.MODE_WRITE)
		{
			Writer wr = (Writer)pipe.getHandle ();
			try { wr.flush (); }
			catch (IOException e) { return -1; }
			return 0;
		}

		return -1;
	}

	/* == arithmetic built-in functions */
	public void sin (Context ctx, String name, Return ret, Argument[] args) 
	{
		ret.setRealValue (Math.sin(args[0].getRealValue()));
	}

	public void cos (Context ctx, String name, Return ret, Argument[] args)
	{
		ret.setRealValue (Math.cos(args[0].getRealValue()));
	}

	public void tan (Context ctx, String name, Return ret, Argument[] args)
	{
		ret.setRealValue (Math.tan(args[0].getRealValue()));
	}

	public void atan (Context ctx, String name, Return ret, Argument[] args) 
	{
		ret.setRealValue (Math.atan(args[0].getRealValue()));
	}

	public void atan2 (Context ctx, String name, Return ret, Argument[] args) 
	{
		double y = args[0].getRealValue();
		double x = args[1].getRealValue();
		ret.setRealValue (Math.atan2(y,x));
	}

	public void log (Context ctx, String name, Return ret, Argument[] args) 
	{
		ret.setRealValue (Math.log(args[0].getRealValue()));
	}

	public void exp (Context ctx, String name, Return ret, Argument[] args) 
	{
		ret.setRealValue (Math.exp(args[0].getRealValue()));
	}

	public void sqrt (Context ctx, String name, Return ret, Argument[] args) 
	{
		ret.setRealValue (Math.sqrt(args[0].getRealValue()));
	}

	public void bfnint (Context ctx, String name, Return ret, Argument[] args) 
	{
		ret.setIntValue (args[0].getIntValue());
	}

	public void rand (Context ctx, String name, Return ret, Argument[] args)
	{
		ret.setRealValue (random.nextDouble ());
	}

	public void srand (Context ctx, String name, Return ret, Argument[] args) 
	{
		long prev_seed = seed;

		seed = (args == null || args.length == 0)?
			System.currentTimeMillis ():
			args[0].getIntValue();

		random.setSeed (seed);
		ret.setIntValue (prev_seed);
	}

	public void systime (Context ctx, String name, Return ret, Argument[] args) 
	{
		long msec = System.currentTimeMillis ();
		ret.setIntValue (msec / 1000);
	}

	public void strftime (Context ctx, String name, Return ret, Argument[] args) throws Exception
	{
		String fmt = (args.length<1)? "%c": args[0].getStringValue();
		long t = (args.length<2)? (System.currentTimeMillis()/1000): args[1].getIntValue();
		ret.setStringValue (strftime (fmt, t));
	}

	public void strfgmtime (Context ctx, String name, Return ret, Argument[] args) throws Exception
	{
		String fmt = (args.length<1)? "%c": args[0].getStringValue();
		long t = (args.length<2)? (System.currentTimeMillis()/1000): args[1].getIntValue();
		ret.setStringValue (strfgmtime (fmt, t));
	}

	/* miscellaneous built-in functions */
	public void system (Context ctx, String name, Return ret, Argument[] args) throws Exception
	{
		ret.setIntValue (system (args[0].getStringValue()));
	}

	/* == utility functions == */
	private Process popen (String command) throws IOException
	{
		String full;

		/* TODO: consider OS names and versions */
		full = System.getenv ("ComSpec");
		if (full != null)
		{	
			full = full + " /c " + command;
		}
		else
		{
			full = System.getenv ("SHELL");
			if (full != null)
			{
				full = "/bin/sh -c \"" + command + "\"";
			}
			else full = command;
		}

		return Runtime.getRuntime().exec (full);
	}
}